### PR TITLE
[chore] set fixed Go patch version to prevent CI failures

### LINKED ISSUE #31213: [ci] `Gen genotelcontribcol` check fails because `cmd/otelcontribcol/go.mod` toolchain line is modified
### Component(s)

cmd/otelcontribcol

### Describe the issue you're reporting

Looks like after updating Contrib to Go 1.21 in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31109, the check on `make genotelcontribcol` fails depending on the version of Go used on the GitHub runner.

The problem (as I understand) is that with Go 1.21, a new [toolchain](https://go.dev/doc/toolchain) feature was added that adds a new line to `go.mod` specifying the version of Go toolchain to use. So the PR #31109 added this line to otelcontribcol's `go.mod`: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/c233bc7b20850fb5febecbf38d6adae3616bc13e/cmd/otelcontribcol/go.mod#L7

The problem is, when the GitHub runner uses Go `v1.21.7` like in [this run](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/7884160170/job/21512668727), it apparently modifies this line in the [Gen geotelcontribcol](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/c233bc7b20850fb5febecbf38d6adae3616bc13e/.github/workflows/build-and-test.yml#L215-L218) check to `toolchain go1.21.7`, which fails the check (the check expects no changes are introduced by running `make genotelcontribcol`).

After the author of the [PR](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31202) that failed the check (the PR contains no changes related to the otelcontribcol's go.mod file) changed this "toolchain" line accordingly to `go1.21.7` in her PR, she got another failure, this time from a GitHub runner that was running Go `1.21.6`. :sweat_smile:

My take is that this could be resolved by pinning the Go version to an exact patch version like `1.21.7` for all builds, instead of the current `~1.21.5`. Any other ideas?

### PR DESCRIPTION
**Pull request by andrzej-stencel**
Fixes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/31213

### CODE REVIEW

**Review by songy23 — APPROVED** [review]

File: exporter/cassandraexporter/go.mod [hunk 1]
```diff
@@ -2,7 +2,7 @@ module github.com/open-telemetry/opentelemetry-collector-contrib/exporter/cassan
 
 go 1.21
 
-toolchain go1.21.6
+toolchain go1.21.7
```
Comment (TylerHelmuth) [inline exporter/cassandraexporter/go.mod:5]: How come this specific exporter needs the toolchain?

Comment (astencel-sumo) [reply exporter/cassandraexporter/go.mod:5]: No idea! :confused:

Comment (TylerHelmuth) [reply exporter/cassandraexporter/go.mod:5]: It was added via https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31109.  @codeboten do you know? If it is removed does stuff break?

Comment (codeboten) [reply exporter/cassandraexporter/go.mod:5]: it's unclear to me why this exporter needs it. I'd suspect one of its dependencies causes this?

**Comment by codeboten** [comment]
@astencel-sumo @TylerHelmuth  i just removed the directive in this PR https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31224, will see if CI is happy with this

File: exporter/cassandraexporter/go.mod (hunk 1 above)
Comment (astencel-sumo) [reply exporter/cassandraexporter/go.mod:5]: Let me delete the line and see what happens :smile:

**Review by TylerHelmuth — APPROVED** [review]

**Pull request merged**